src/common/droid-util.c

```diff
@@ -1154,7 +1153,76 @@ static bool stream_config_fill(audio_devices_t devices,
         sample_spec->channels = 1;
         hal_channel_mask = AUDIO_CHANNEL_IN_MONO;
 #endif
+        pa_log_debug("Fill config: Use %d channel(s) for voice call.", sample_spec->channels);
+    }
+
+    if (!pa_convert_format(sample_spec->format, CONV_FROM_PA, &hal_audio_format)) {
+        pa_log("Sample spec format %u not supported.", sample_spec->format);
+        goto fail;
+    }
+
+    /* As input sample metrics are based on user request we need to make sure that the sample
+     * format requested is actually usable with the input route. */
+    if (!output) {
+        const pa_droid_config_device *ddef;
+        uint32_t tmp;
+        const char *fmt;
+
+        pa_log_debug("Fill config: Try to use format %s, sample rate %u",
+            pa_string_convert_num_to_str(CONV_STRING_FORMAT, hal_audio_format, &fmt) ? fmt : "<unknown>",
+            sample_spec->rate);
+
+        SLLIST_FOREACH(ddef, device_def) {
+            if ((ddef->formats & hal_audio_format) &&
+                (ddef->devices & devices)) {
+
+                for (i = 0; ddef->sampling_rates[i]; i++) {
+                    if (ddef->sampling_rates[i] == (uint32_t) -1 ||
+                        ddef->sampling_rates[i] == sample_spec->rate) {
+                        pa_log_debug("Fill config: Using requested sample rate %d", sample_spec->rate);
+                        goto format_found;
+                    }
+                }
+            }
+        }
+
+        SLLIST_FOREACH(ddef, device_def) {
+            if (!(ddef->devices & devices) ||
+                !(ddef->formats & AUDIO_FORMAT_PCM_16_BIT))
+                continue;
+
+            hal_audio_format = AUDIO_FORMAT_PCM_16_BIT;
+            pa_assert_se(pa_convert_format(hal_audio_format, CONV_FROM_HAL, &tmp));
+            sample_spec->format = tmp;
+
+            for (i = 0; ddef->sampling_rates[i]; i++) {
+                if (ddef->sampling_rates[i] != (uint32_t) -1 &&
+                    ddef->sampling_rates[i] != sample_spec->rate)
+                    continue;
+
+                pa_log_debug("Fill config: Override sample format, use %s.",
+                    pa_string_convert_num_to_str(CONV_STRING_FORMAT, hal_audio_format, &fmt) ? fmt : "<unknown>");
+                goto format_found;
+            }
+
+            /* Start from highest going to lowest */
+            for (i = 0; ddef->sampling_rates[i+1]; i++)
```

Comment:
(theoretical, but handled elsewhere?) empty array -> [0 + 1] goes out of bounds

Reply:
Updated and now empty array is checked in the beginnig of that block.